src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -641,7 +641,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
         // A module reference in a pattern has type Foo.type, not "object Foo"
         narrowIf(checkStable(tree), sym.isModuleNotMethod)
       else if (isModuleTypedExpr)                     // (3)
-        narrowIf(tree, condition = true)
+        narrowIf(tree, condition = !sym.isConstant)
```

Comment:
Maybe `!isConstantType(tree.tpe)` is somewhat more direct, as `narrowIf` is calling `tree.setType`?

Reply:
You're probably right. I forget that `tree.symbol.tpe` is different from `tree.tpe`, so I'm not sure why I reached for the symbol test, except that first I added the condition to `isModuleTypedExpr`. `#3` is about the tree type.